### ISSUE TITLE
Homework Announcement

### ISSUE THREAD

**Issue by markshin** [snu-sf-class/swpp201801#13]
[swpp2018_hw.pdf](https://github.com/snu-sf-class/swpp201801/files/1863294/swpp2018_hw.pdf)

**Comment by joks1997** [comment]
초기 환경 세팅에서 /home/swpp/ 경로를 사용하라고 해주셨는데 mac OS 사용자들은 home 디렉토리가 보호되어 있어 파일을 쓸 수 없습니다. 다른 경로에서 작업할 수 있도록 해주실 수 있는지 문의 드립니다.

**Comment by wogus20002** [comment]
제가 우분투가 익숙치 않아서 그런데 adduser 로 swpp 라는 계정을 꼭 새로 만들어야 되나요? 아니면 그냥 home/swpp 폴더만 만들면 되나요?

adduser 로 새로 계정을 만드니까 모든 프로그램들을 다시 설치해야 되고 sudo 권한 문제도 계속 발생합니다.

**Comment by chongkong** [comment]
경로를 꼭 `/home/swpp/homework2`로 맞추어야 한다는 것은 제출 시 virtualenv 디렉토리까지 zip파일로 압축하고 해당 virtualenv의 python을 이용해서 채점을 하시려는 것처럼 생각되는데요, 그보다는 채점에 사용할 인터프리터 버전 및 인스톨된 패키지들을 명시해주시고 채점용 인터프리터를 사용하는게 좋을 것 같다고 생각됩니다.

**Comment by chongkong** [comment]
참고로 virtualenv로 생성된 activate 스크립트에서는 `VIRTUAL_ENV` 라는 환경변수에 virtualenv 절대경로를 주입하기 때문에 경로가 바뀌면 사용할 수 없습니다.

``` bash
# ...
VIRTUAL_ENV="/Users/home/homeworktwo/env"
export VIRTUAL_ENV

_OLD_VIRTUAL_PATH="$PATH"
PATH="$VIRTUAL_ENV/bin:$PATH"
export PATH
# ...
```

**Comment by ryanking13** [comment]
저도 @chongkong 님처럼 생각했는데, 디스크립션을 보면 제출할때 생성한 virtualenv는 제출하지 않는 것 으로 보이네요. 그래서 어떠한 이유로 정해진 경로를 사용해야하는 지 모르겠습니다.

**Comment by chongkong** [comment]
또 promises DELETE 에 대한 설명이 덜 작성되어 있는 것 같습니다.

![image](https://user-images.githubusercontent.com/4899492/38133027-6a67eabe-3448-11e8-812d-c4bc22142d32.png)

**Comment by markshin** [comment]
@chongkong 오타를 수정했습니다. 죄송합니다.

**Comment by markshin** [comment]
@joks1997 우분투를 사용해서 숙제를 하셨으면 합니다.

**Comment by petrosyh** [comment]
@chongkong @ryanking13 venv를 사용한 테스트를 사용해려고해서 절대주소를 강제한 것이 맞습니다.
venv를 포함하지 않게 제출하라고 한 것은 실수이니, 그것을 다시 수정해서 올리도록 하겠습니다.

**Comment by petrosyh** [comment]
@chongkong @ryanking13 확인한 후에 다시 답변드리겠습니다.

**Comment by petrosyh** [comment]
@wogus20002 새로 계정을 만들고 세팅을 해서 사용하시면 좋을 것 같습니다.

**Comment by petrosyh** [comment]
@chongkong @ryanking13

**Comment by sanichdaniel** [comment]
mac유저면 우분투를 사용하는것 이외에는 다른 발법이 없는건가요??

**Comment by petrosyh** [comment]
@sanichdaniel 예, 일단 그러합니다. Virtual Box는 mac os 에서도 잘 작동되니 그 위에서 진행하는 것을 권장합니다.

**Comment by kajebiii** [comment]
1. """또한 약속 당사자가 아닌 유저는 promises/id url 에 접근할 수 없다."""에서 약속 당사자는 user1을 의미하나요? 아니면 user1 또는 user2을 의미하나요?

2. """promises/id/ url 의 put method 를 통해서 약속 관계를 수정할 때도 sinceWhen이 tilWhen 보다 이른 시간이어야 한다(400 bad request 로 처리)."""
만 조건으로 써져 있는데 이때는 (같은 id 와의 약속을 만들 수 없다)의 조건은 생각하지 않아도 되는 것인가요?

**Comment by petrosyh** [comment]
@ryanking13 @chongkong 
일단, 작년 채점 스크립트와 함께 새로운 스크립트를 만드는 중입니다만, 경로를 강제할 필요는 없을 것 같습니다. 새로운 issue를 통해서 공지드리겠습니다.

**Comment by petrosyh** [comment]
@kajebiii 
1. user1, user2 전부입니다.

2. 두 조건중에(sinceWhen < tilWhen과 user가 같으면 안되는 조건) 하나라도 충족하지 않으면 400을 띄우면 좋겠습니다.

**Comment by kj4569** [comment]
안녕하세요! 아직 모르는 것이 많아서 질문드립니다. 

1. promises의 경우 id = 1, 2, 3,.., n개를 만들고 삭제한 후 다시 만들면 id = n+1부터 만들어지던데 상관없을까요? 완전히 초기화를 시키는 방법이 따로 있나요?

2. sinceWhen/tilWhen의 경우 "2017-01-01T01:00:00"와 같이 출력시켜도 되나요? created은 "2018-04-02T21:40:57.316930"와 같이 출력되는데 왜 둘이 형식이 다르게 나오는지 모르겠네요ㅠㅠ

감사합니다!

**Comment by petrosyh** [comment]
@kj4569 
1. 딱히 상관없을 것 같습니다. 그냥 db와 migration을 다 날려버린 후에 makemigration/migrate를 다시 한 후에도 그런가요?

2. 제 생각에는 2018-04-02T21:40:57.316930Z처럼 포멧을 지키는 쪽이 안전할 것 같습니다.

3. 쉘 내에서 띄우면 충분할 것 같습니다.

**Comment by kj4569** [comment]
@petrosyh 
말씀해주신대로 했더니 1번은 해결됐습니다. 감사합니다!

**Comment by kjp4155** [comment]
인증되지 않거나 다른(제 3자) 유저가 약속내용에 접근하거나 수정/삭제하려는 경우 이를 막아야 한다고 되어 있는데요, 단순히 400 에러를 띄우는 것으로 충분할까요?

**Comment by petrosyh** [comment]
@kjp4155 예 그렇게 해주시면 될 것 같습니다.

**Comment by ilil96** [comment]
안녕하세요, 기술적인 질문을 해도 되는지 모르겠지만 감이 잡히지 않아 질문합니다.
promises/ 에서 User 클래스(django에서 기본으로 제공하는 클래스)가 is not json serializable 하다는 type error가 뜹니다.
User 클래스를 serializable하게 하려면 어떻게 해야할까요?
감사합니다.

**Comment by jeuxdeau** [comment]
채점 스크립트와 과제 문서를 보면 PUT을 할 때 sinceWhen과 tilWhen만 입력받는 것 같은데요, user가 같은지를 체크해주어야 할 필요가 있나요?

**Comment by ilil96** [comment]
promises_as_inviter와 promises_as_invitee는 primaryrelatedfield를 통해서 받을 수 있는데,
whole_promises는 어떻게 받아야 하나요?
감사합니다.

**Comment by petrosyh** [comment]
@jeuxdeau user가 같은지 체크해준다고 하시는 말이 무슨 말씀이신지 모르겠습니다. 조금 더 구체적으로 질문 부탁드립니다. 또한, 약속 당사자가 아닌 사람은 promises/id/ url에 접근할 수 없다고 했으니, 당사자가 아닌 user들이 접근하지 못하게 권한을 주는 정도로 충분할 것 같습니다.

@ilil96 whole promises를 어떤식으로 구성하는지는 학생분께서 자유롭게 숙제 스펙에 맞게만 구현하시면 됩니다. 다만 힌트를 드리자면 promises_as_inviter와 promises_as_invitee의 합집합이 whole_promises라는 것을 숙제 스펙을 보시면 유추하실 수 있을 것입니다.

**Comment by jeuxdeau** [comment]
@petrosyh 위에 조교님과 kajebiii님과의 대화에서, put에서도 user 둘이 같은지를 체크해 달라고 말씀하셔서요.

질문을 좀 더 확장하면, put에서 user를 받거나 post에서 user1을 받는 것처럼 필수가 아닌 입력을 받는 것까지 고려해야 하나요?

**Comment by sunnyday0208** [comment]
@petrosyh  저도 ilil96님이 질문하신 것에서 헤매고 있는데, manyrelatedfield 두 개를 합치는 방법을 아무리 찾아봐도 모르겠습니다ㅠㅠ

**Comment by minkiminki** [comment]
@jeuxdeau 과제 스펙에 주어진 입력 포맷만으로 테스트할 것입니다.
말씀하신대로 promise 수정에서 user는 입력 안 받으므로 user1과 user2랑 같은지 생각할 필요 없는게 맞습니다.
